pkg/controller/scyllacluster/migrate.go

```diff
@@ -0,0 +1,431 @@
+// Copyright (c) 2024 ScyllaDB.
+
+package scyllacluster
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"maps"
+	"strings"
+
+	scyllav1 "github.com/scylladb/scylla-operator/pkg/api/scylla/v1"
+	scyllav1alpha1 "github.com/scylladb/scylla-operator/pkg/api/scylla/v1alpha1"
+	"github.com/scylladb/scylla-operator/pkg/controllerhelpers"
+	"github.com/scylladb/scylla-operator/pkg/helpers/slices"
+	"github.com/scylladb/scylla-operator/pkg/internalapi"
+	"github.com/scylladb/scylla-operator/pkg/naming"
+	"github.com/scylladb/scylla-operator/pkg/pointer"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+)
+
+func MigrateV1ScyllaClusterSpecToV1Alpha1ScyllaDBDatacenterSpec(scName string, scSpec scyllav1.ScyllaClusterSpec) (scyllav1alpha1.ScyllaDBDatacenterSpec, error) {
+	var migrateErrs []error
+
+	return scyllav1alpha1.ScyllaDBDatacenterSpec{
+		Metadata: func() scyllav1alpha1.ObjectTemplateMetadata {
+			if scSpec.PodMetadata == nil {
+				return scyllav1alpha1.ObjectTemplateMetadata{}
+			}
+			return scyllav1alpha1.ObjectTemplateMetadata{
+				Labels:      scSpec.PodMetadata.Labels,
+				Annotations: scSpec.PodMetadata.Annotations,
+			}
+		}(),
+		ClusterName:    scName,
+		DatacenterName: scSpec.Datacenter.Name,
+		ScyllaDB: scyllav1alpha1.ScyllaDB{
+			// v1.ScyllaCluster doesn't provide API to specify cluster/datacenter wide properties of scylladb.
+			ScyllaDBTemplate: scyllav1alpha1.ScyllaDBTemplate{},
+			Image:            fmt.Sprintf("%s:%s", scSpec.Repository, scSpec.Version),
```

Comment:
version can be empty string which will make this and invalid reference ending with `:`

Reply:
added error handling